File: src/crewlet/env_file.py
"""THE ``.env`` assignment grammar — one writer for two provisioners.

A provisioning run mints credentials that are unretrievable afterwards,
and an env file is where several of the CLIs put them.  What that file
must guarantee is round-tripping: the value the engine reads back, and
the value a shell gets from ``source``, have to be the value that was
minted.  Two very ordinary characters break that when a value is
written bare — a space makes ``source`` fail on the line (and, because
``source`` stops there, silently abandons every credential BELOW it),
and ``${…}`` is interpolated away by python-dotenv in the reader the
engine actually uses.

This module is that grammar, in one place, because it was in two: the
Slack provisioner reasoned it through and refused the shapes it could
not represent, while the shared ``TokenSink`` env-file backend wrote
``f"{var}={token}"`` and hoped.  Both write secrets, so a divergence
means the safety of a credential depends on which CLI minted it.

It imports nothing from Crewlet, like :mod:`crewlet.env_refs` and
:mod:`crewlet.queue.topics`, so every writer can depend on it —
and ``tests/test_provisioning/test_env_file.py`` fails the build on a
second definition of the grammar.
"""

from __future__ import annotations

import re

#: Matches ``KEY=…`` / ``export KEY=…`` assignment lines.
ASSIGNMENT_RE = re.compile(r"^\s*(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)\s*=")

#: Characters that force quoting.  ``$`` is included because python-dotenv
#: interpolates ``${VAR}`` inside unquoted and double-quoted values —
#: only single quotes disable it, so ``$`` routes to the single-quoted
#: form below.
_NEEDS_QUOTING_RE = re.compile(r"[\s#'\"\\$]")

_EXPORT_PREFIX = "export "


class EnvValueError(ValueError):
    """A value with no faithful ``.env`` representation."""


def format_assignment(key: str, value: str, *, export: bool = False) -> str:
    """Render one ``KEY=value`` line that reads back verbatim.

    Forms, chosen by content:

    - plain ``KEY=value`` when no character needs quoting;
    - ``KEY='value'`` when the value contains a bare ``$`` (quoting
      keeps it out of the double-quote escape rules);
    - ``KEY="value"`` with backslash escapes otherwise.

    ``export`` prefixes the line, which is the shape ``--print`` emits
    so pasted output round-trips; every quoting form above stays valid
    after it.

    Two value shapes have NO faithful python-dotenv representation and
    are refused loudly rather than persisted as something the engine
    would read back differently: ``${…}`` sequences (python-dotenv
    interpolates them in EVERY quoting form, single quotes included),
    and a bare ``$`` combined with a single quote (the only safe quoting
    for ``$`` can't contain ``'``).  Real credentials are URL-safe
    strings, so neither shape occurs in practice — but a refusal an
    operator can read beats a token that silently changes on its way
    back out.
    """
    prefix = _EXPORT_PREFIX if export else ""
    if value and not _NEEDS_QUOTING_RE.search(value):
        return f"{prefix}{key}={value}"
    if "${" in value:
        msg = (
            f"cannot faithfully store {key}: python-dotenv interpolates "
            "${...} sequences in every quoting form, so the value would "
            "read back changed"
        )
        raise EnvValueError(msg)
    if "$" in value:
        if "'" in value:
            msg = (
                f"cannot faithfully store {key}: the value combines '$' "
                "and a single quote, which python-dotenv cannot round-trip"
            )
            raise EnvValueError(msg)
        return f"{prefix}{key}='{value}'"
    escaped = value.replace("\\", "\\\\").replace('"', '\\"')
    return f'{prefix}{key}="{escaped}"'


def parse_assignment(line: str) -> tuple[str, str] | None:
    """Split one assignment line into ``(key, value)``, or ``None``.

    The read half of :func:`format_assignment`, and deliberately its
    exact inverse: it unwraps either quoting form and undoes the
    double-quoted escapes.  A reader that only stripped the quotes
    would hand back ``a\\"b`` for a value that was written as ``a"b``.
    """
    match = ASSIGNMENT_RE.match(line)
    if match is None:
        return None
    key = match.group(1)
    raw = line.split("=", 1)[1].strip()
    if len(raw) >= 2 and raw[0] == raw[-1] and raw[0] in "'\"":
        inner = raw[1:-1]
        if raw[0] == '"':
            inner = inner.replace('\\"', '"').replace("\\\\", "\\")
        return key, inner
    return key, raw


def is_exported(line: str) -> bool:
    """Whether an assignment line carries the ``export`` prefix."""
    return line.strip().startswith(_EXPORT_PREFIX)


__all__ = [
    "ASSIGNMENT_RE",
    "EnvValueError",
    "format_assignment",
    "is_exported",
    "parse_assignment",
]

File: internal/envfile/envfile.go
// Package envfile is THE grammar of one `.env` assignment.
//
// # Why a package for one line of text
//
// A credential file has TWO readers and they do not agree, so a line that
// satisfies one can silently break the other:
//
//   - an OPERATOR reads it with `source .env`, where a bare space ends the
//     assignment and the rest of the line becomes a command — and the shell
//     then abandons every credential BELOW it in the file;
//   - a DOTENV TOOL reads it — direnv, docker compose's env_file, a CI
//     step's loader — and expands ${...} inside unquoted and double-quoted
//     values, so a token containing one is silently rewritten.
//
// The first failure is the one that hurts: it is silent, it is positional,
// and the symptom is a company where the seats defined after some particular
// token stop authenticating.
//
// # The ENGINE is not one of the readers
//
// It reads `${VAR}` from this node's secret store and then from the PROCESS
// environment, and from nowhere else — so a file written here has to be
// sourced, or fed to the process some other way, before the engine starts.
// A dotenv loader in the engine would be a third source of truth for
// secrets, discovered by filename, able to override the Tier A keyring that
// opens the store: exactly the inversion the two-tier design refuses. The
// path that needs no file at all is `-secret-store`.
//
// Every provisioner that mints a credential writes one of these lines. When
// each built its own, whether a minted token survived depended on which CLI
// minted it — one reasoned the quoting through and refused what it could not
// represent, another wrote a bare fmt.Sprintf("%s=%s"). This package is the
// one answer, and the suite fails the build on a new hand-built assignment.
//
// It imports nothing from crewlet, so even the config loader can use it.
package envfile

import (
	"fmt"
	"regexp"
	"strings"
)

// AssignmentRE matches one assignment line, with or without `export`.
//
// The `export` prefix is accepted because an operator's file usually has it —
// it is what makes `source .env` put the values into the CHILD processes an
// operator then starts — and a parser that rejected it would rewrite a file
// it could not read.
var AssignmentRE = regexp.MustCompile(
	`^\s*(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)\s*=(.*)$`)

// ErrUnrepresentable reports a value no assignment line can carry safely.
//
// Its own error because the caller's answer is specific: a provisioner that
// minted a token it cannot write must REVOKE it rather than leave it
// standing, since a credential that exists and is not recorded is one nobody
// can use and nobody will remember to remove.
var ErrUnrepresentable = fmt.Errorf("envfile: the value cannot be written safely")

// FormatAssignment renders NAME=value so that BOTH readers agree.
//
// # Single quotes, and why not double
//
// A single-quoted shell string is literal: no expansion, no escapes, no
// interpretation of $ or ` or \. A double-quoted one expands $NAME and
// ${NAME} in both the shell AND the dotenv loader — so a token containing a
// literal `$` (which base64 and many vendor tokens do) would reach the
// engine with a chunk of itself replaced by an environment variable, or by
// nothing.
//
// The only thing a single-quoted string cannot contain is a single quote.
// Rather than the usual '\” splice — which `source` handles and several
// dotenv readers do not — such a value is REFUSED. A provisioner can then
// mint a different one; silently writing something one reader mangles is
// how a credential comes to differ between the two.
func FormatAssignment(name, value string) (string, error) {
	if !NameOK(name) {
		return "", fmt.Errorf("envfile: %q is not a variable name", name)
	}
	if strings.ContainsAny(value, "'\n\r") {
		// A NEWLINE IS AS FATAL AS A QUOTE and for the same reason: the
		// second line is not an assignment, so `source` runs it.
		return "", fmt.Errorf("%w: %s contains a single quote or a newline",
			ErrUnrepresentable, name)
	}
	return "export " + name + "='" + value + "'", nil
}

// NameOK reports whether a name is a usable variable name.
func NameOK(name string) bool {
	if name == "" {
		return false
	}
	for i, r := range name {
		switch {
		case r == '_' || (r >= 'A' && r <= 'Z') || (r >= 'a' && r <= 'z'):
		case i > 0 && r >= '0' && r <= '9':
		default:
			return false
		}
	}
	return true
}

// ParseAssignment reads one line back, reporting whether it was one.
//
// It UNDERSTANDS all three quoting forms rather than only the one
// FormatAssignment writes, because the file it reads was usually written by
// a person: an operator's hand-written `TOKEN=abc` must be recognised as the
// same variable, or a rotation appends a second assignment and which one
// wins depends on the reader.
func ParseAssignment(line string) (name, value string, ok bool) {
	m := AssignmentRE.FindStringSubmatch(line)
	if m == nil {
		return "", "", false
	}
	return m[1], unquote(strings.TrimSpace(m[2])), true
}

// unquote strips one layer of matching quotes.
//
// A BARE value is trimmed of trailing comment-like text only when it is
// quoted, never when it is not: `TOKEN=abc # note` is a shell assignment of
// `abc` with a comment, but a dotenv reader disagrees about it — so this
// takes the whole remainder, which is what the engine's loader does, and
// leaves the ambiguity visible rather than resolving it two ways.
func unquote(value string) string {
	if len(value) < 2 {
		return value
	}
	first, last := value[0], value[len(value)-1]
	if first == last && (first == '\'' || first == '"') {
		return value[1 : len(value)-1]
	}
	return value
}
